source/s3_parallel_input_stream.c

```diff
@@ -0,0 +1,165 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include "aws/s3/private/s3_parallel_input_stream.h"
+
+#include <aws/common/atomics.h>
+#include <aws/common/file.h>
+#include <aws/common/string.h>
+#include <aws/common/task_scheduler.h>
+
+#include <aws/io/event_loop.h>
+#include <aws/io/future.h>
+#include <aws/io/stream.h>
+
+#include <errno.h>
+#include <sys/stat.h>
+
+void aws_parallel_input_stream_init_base(
+    struct aws_parallel_input_stream *stream,
+    struct aws_allocator *alloc,
+    const struct aws_parallel_input_stream_vtable *vtable,
+    void *impl) {
+
+    AWS_ZERO_STRUCT(*stream);
+    stream->alloc = alloc;
+    stream->vtable = vtable;
+    stream->impl = impl;
+    aws_ref_count_init(&stream->ref_count, stream, (aws_simple_completion_callback *)vtable->destroy);
+}
+
+struct aws_parallel_input_stream *aws_parallel_input_stream_acquire(struct aws_parallel_input_stream *stream) {
+    if (stream != NULL) {
+        aws_ref_count_acquire(&stream->ref_count);
+    }
+    return stream;
+}
+
+struct aws_parallel_input_stream *aws_parallel_input_stream_release(struct aws_parallel_input_stream *stream) {
+    if (stream != NULL) {
+        aws_ref_count_release(&stream->ref_count);
+    }
+    return stream;
+}
+
+struct aws_future_bool *aws_parallel_input_stream_read(
+    struct aws_parallel_input_stream *stream,
+    uint64_t offset,
+    struct aws_byte_buf *dest) {
+    /* Ensure the buffer has space available */
+    if (dest->len == dest->capacity) {
+        struct aws_future_bool *future = aws_future_bool_new(stream->alloc);
+        aws_future_bool_set_error(future, AWS_ERROR_SHORT_BUFFER);
+        return future;
+    }
+    /* TODO: restore the buffer on failure. */
+    struct aws_future_bool *future = stream->vtable->read(stream, offset, dest);
+    return future;
+}
+
+struct aws_parallel_input_stream_from_file_impl {
+    struct aws_parallel_input_stream base;
+
+    struct aws_string *file_path;
+    uint64_t last_modified_time;
```

Comment:
highly debatable: ensuring that the file remains unchanged is .... honorable. But like ... that's user error man. And I get the feeling supporting this cross-platform could be annoying

Unless Java says they do this, because of some real reason, I'd just keep it simple and delete this
```suggestion
```

Reply:
They do check the file last modified time. https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncRequestBody.java#L397

I added it because I saw them doing it. I can check with them if there are real reasons behind.

Reply:
well, checked the PR added it. https://github.com/aws/aws-sdk-java-v2/pull/2796

They don't have a real ask for this, was trying to be nice to give more info about checksum mismatch

Reply:
ok, fair